zabapgit.prog.abap

```diff
@@ -11476,6 +11476,344 @@ CLASS lcl_object_sfbs IMPLEMENTATION.
+    DATA lt_w3html   TYPE STANDARD TABLE OF w3html.
+    DATA lt_w3params TYPE STANDARD TABLE OF wwwparams.
+    DATA l_xstring   TYPE xstring.
+    DATA ls_wwwparam TYPE wwwparams.
+    DATA l_size      TYPE int4.
+    DATA l_base64str TYPE string.
+    DATA lo_utility  TYPE REF TO cl_http_utility.
+
+    ls_key-relid = ms_item-obj_type+2(2).
+    ls_key-objid = ms_item-obj_name.
+
+    SELECT SINGLE * INTO CORRESPONDING FIELDS OF ls_key
+      FROM wwwdata
+      WHERE relid = ls_key-relid
+      AND   objid = ls_key-objid
+      AND   srtf2 = 0.
+
+    IF sy-subrc IS NOT INITIAL.
+      RETURN.
+    ENDIF.
+
+    CALL FUNCTION 'WWWDATA_IMPORT'
+      EXPORTING  key               = ls_key
+      TABLES     mime              = lt_w3mime
+                 html              = lt_w3html
+      EXCEPTIONS wrong_object_type = 1
+                 import_error      = 2.
+
+    IF sy-subrc IS NOT INITIAL.
+      _raise 'Cannot read W3xx data'.
+    ENDIF.
+
+    CALL FUNCTION 'WWWPARAMS_READ_ALL'
+      EXPORTING   type             = ls_key-relid
+                  objid            = ls_key-objid
+      TABLES      params           = lt_w3params
+      EXCEPTIONS  entry_not_exists = 1.
+
+    IF sy-subrc IS NOT INITIAL.
+      _raise 'Cannot read W3xx data'.
+    ENDIF.
+
+    READ TABLE lt_w3params INTO ls_wwwparam WITH KEY name = 'filesize'.
+    IF sy-subrc IS NOT INITIAL.
+      _raise 'Cannot read W3xx filesize'.
+    ENDIF.
+
+    l_size = ls_wwwparam-value.
+
+    CASE ls_key-relid.
+      WHEN 'MI'.
+        CALL FUNCTION 'SCMS_BINARY_TO_XSTRING'
+          EXPORTING  input_length = l_size
+          IMPORTING  buffer       = l_xstring
+          TABLES     binary_tab   = lt_w3mime
+          EXCEPTIONS failed       = 1.
+      WHEN 'HT'.
+        CALL FUNCTION 'SCMS_TEXT_TO_XSTRING'
+          IMPORTING  buffer       = l_xstring
+          TABLES     text_tab     = lt_w3html
+          EXCEPTIONS failed       = 1.
+      WHEN OTHERS.
+        _raise 'Wrong W3xx type'.
+    ENDCASE.
+
+    IF sy-subrc IS NOT INITIAL.
+      _raise 'Cannot convert W3xx to xstring'.
+    ENDIF.
+
+    CREATE OBJECT lo_utility.
+    l_base64str = lo_utility->encode_x_base64( unencoded = l_xstring ).
+
+    io_xml->add( iv_name = 'NAME'
+                 ig_data = ls_key-objid ).
+
+    io_xml->add( iv_name = 'TEXT'
+                 ig_data = ls_key-text ).
+
+    io_xml->add( iv_name = 'DATA'
+                 ig_data = l_base64str ).
+
+    " TODO normal params serialization
+
+    DEFINE w3mi_save_param.
+      READ TABLE lt_w3params INTO ls_wwwparam WITH KEY name = &1.
+      IF sy-subrc IS NOT INITIAL.
+        _raise 'Cannot read required W3xx param'.
+      ENDIF.
+      io_xml->add( iv_name = &1
+                   ig_data = ls_wwwparam-value ).
+    END-OF-DEFINITION.
+
+    w3mi_save_param 'fileextension'.
+    w3mi_save_param 'filename'.
+    w3mi_save_param 'filesize'.
+    w3mi_save_param 'mimetype'.
+    w3mi_save_param 'version'.
```

Comment:
Awesome! I like this style :) Nice and clean code now :) Is it this New XML feature closed recently ? Does it also supports attributes in some way ?